### PR TITLE
ws: Add cockpit-ws command line options

### PR DESCRIPTION
**Pull request by petervo**
Adds the following options to cockpit-ws

 * address: The address to bind to
 * static-dir: The preferred location for static files
 * config-dir: The location for the config file and certificates

### CODE REVIEW

File: doc/man/cockpit-ws.xml [hunk 1]
```diff
@@ -132,6 +135,39 @@ $ sudo remotectl certificate
         </listitem>
       </varlistentry>
       <varlistentry>
+        <term><option>--address</option> <replaceable>ADDRESS</replaceable></term>
+        <listitem>
+          <para>
+            Bind to address <replaceable>ADDRESS</replaceable> instead of binding to
+            all available addresses. Usually Cockpit is started on demand by
+            <command>systemd</command> socket activation, and this option has no effect.
+            Update the <literal>ListenStream</literal> directive <filename>cockpit.socket</filename>
```
Comment (stefwalter) [inline doc/man/cockpit-ws.xml]: "In that case, update the"...

File: src/common/cockpitcertificate.c [hunk 2]
```diff
@@ -259,7 +260,7 @@ cockpit_certificate_locate (gboolean create_if_necessary,
                             GError **error)
 {
   gchar *cert_path = NULL;
-  const gchar *cert_dir = PACKAGE_SYSCONF_DIR "/cockpit/ws-certs.d";
+  gchar *cert_dir = g_build_filename (cockpit_conf_get_dir(), "ws-certs.d", NULL);
```
Comment (stefwalter) [inline src/common/cockpitcertificate.c]: Missing space after function name.

File: src/common/cockpitconf.c [hunk 3]
```diff
@@ -78,28 +79,47 @@ void
...
+    dir = g_path_get_dirname (cockpit_config_file);
+
+  /* only add cockpit_config_dir if cockpit_config_file doesn't
+   * already have a directory
+   */
+  if (cockpit_config_dir && g_strcmp0 (dir, ".") == 0)
+    file = g_build_filename (cockpit_config_dir, cockpit_config_file, NULL);
+  else if (cockpit_config_file)
+    file = g_strdup (cockpit_config_file);
+
```
Comment (stefwalter) [inline src/common/cockpitconf.c]: Extra empty lines?

File: src/common/cockpitwebserver.c [hunk 4]
```diff
@@ -241,13 +245,24 @@ cockpit_web_server_set_property (GObject *object,
...
       server->port = g_value_get_int (value);
       break;
 
+    case PROP_ADDRESS:
+      address = g_value_get_string (value);
+      if (address)
+        {
+          server->address = g_inet_address_new_from_string (address);
+          if (!server->address)
+            g_warning ("Couldn't parse IP address from: %s", address);
```
Comment (stefwalter) [inline src/common/cockpitwebserver.c]: "Couldn't parse address to listen on: %s"

File: src/remotectl/certificate.c [hunk 5]
```diff
@@ -163,9 +164,11 @@ cockpit_remotectl_certificate (int argc,
   gchar *selinux = NULL;
   gchar *group = NULL;
   gchar *user = NULL;
+  gchar *opt_config_dir = NULL;
   int ret = 1;
 
   const GOptionEntry options[] = {
+    {"config-dir", 'c', 0, G_OPTION_ARG_STRING, &opt_config_dir, "Configuration directory defaults to " PACKAGE_SYSCONF_DIR "/cockpit/ if unset)", NULL},
```
Comment (stefwalter) [inline src/remotectl/certificate.c]: Needs space before string, and wrapped to different lines.

File: src/common/cockpitwebserver.c [hunk 6]
```diff
@@ -1330,8 +1357,27 @@ cockpit_web_server_initable_init (GInitable *initable,
     }
   else
     {
+      if (server->address)
+        {
+          socket_address = g_inet_socket_address_new (server->address, server->port);
+          if (socket_address)
+            {
+              failed = !g_socket_listener_add_address (G_SOCKET_LISTENER (server->socket_service),
```
Comment (stefwalter) [inline src/common/cockpitwebserver.c]: Need a unit test for this code path. Or am I missing something?

File: src/common/cockpitwebserver.c (hunk 4 above)
Comment (stefwalter) [inline src/common/cockpitwebserver.c]: Need a unit test for this code path.

File: src/common/cockpitconf.c [hunk 7]
```diff
@@ -78,28 +79,47 @@ void
...
                                         (GDestroyNotify)g_strfreev);
 
+
   if (cockpit_config_file)
+    dir = g_path_get_dirname (cockpit_config_file);
+
+  /* only add cockpit_config_dir if cockpit_config_file doesn't
+   * already have a directory
+   */
+  if (cockpit_config_dir && g_strcmp0 (dir, ".") == 0)
```
Comment (stefwalter) [inline src/common/cockpitconf.c]: Need a unit test for the new cockpit_config_dir stuff.

File: src/common/cockpitconf.c (hunk 3 above)
Comment (petervo) [inline src/common/cockpitconf.c]: fixed

File: src/common/cockpitwebserver.c (hunk 4 above)
Comment (petervo) [inline src/common/cockpitwebserver.c]: added

File: src/common/cockpitwebserver.c (hunk 6 above)
Comment (petervo) [inline src/common/cockpitwebserver.c]: added

File: src/common/cockpitconf.c (hunk 7 above)
Comment (petervo) [inline src/common/cockpitconf.c]: added

**Comment by stefwalter** [comment]
What about using the top two commits here? https://github.com/stefwalter/cockpit/commits/ws-options

**Comment by petervo** [comment]
Seems good, but guess we need to implement the environment arg.

**Comment by stefwalter** [comment]
Here's another fix that I think is necessary: https://github.com/stefwalter/cockpit/commit/0ea00b0cf0f6fcee0254c241f533e29684e4e9d9

**Pull request closed without merging**